openupgradelib/openupgrade_160.py

```diff
@@ -45,3 +58,346 @@ def migrate_translations_to_jsonb(env, fields_spec):
+        Raw HTML fragment converted.
+    """
+    if not html_string:
+        return html_string
+    try:
+        return convert_html_fragment(
+            html_string,
+            ALL_REPLACEMENTS,
+            pretty_print,
+        )
+    except Exception:
+        logger.error("Error converting string BS4 to BS5:\n%s" % html_string)
+        raise
+
+
+def convert_field_bootstrap_4to5(
+    env, model_name, field_name, domain=None, method="orm"
+):
+    """This converts all the values for the given model and field, being
+    able to restrict to a domain of affected records.
+
+    :param env: Odoo environment.
+    :param model_name: Name of the model that contains the field.
+    :param field_name: Name of the field that contains the BS3 HTML content.
+    :param domain: Optional domain for filtering records in the model
+    :param method: 'orm' (default) for using ORM; 'sql' for avoiding problems
+        with extra triggers in ORM.
+    """
+    assert method in {"orm", "sql"}
+    if method == "orm":
+        return _convert_field_bootstrap_4to5_orm(
+            env,
+            model_name,
+            field_name,
+            domain,
+        )
+    records = env[model_name].search(domain or [])
+    return _convert_field_bootstrap_4to5_sql(
+        env.cr,
+        records._table,
+        field_name,
+    )
+
+
+def _convert_field_bootstrap_4to5_orm(env, model_name, field_name, domain=None):
+    """Convert a field from Bootstrap 4 to 5, using Odoo ORM.
+
+    :param odoo.api.Environment env: Environment to use.
+    :param str model_name: Model to update.
+    :param str field_name: Field to convert in that model.
+    :param domain list: Domain to restrict conversion.
+    """
+    domain = domain or [(field_name, "!=", False), (field_name, "!=", "<p><br></p>")]
+    records = env[model_name].search(domain)
+    update_field_multilang(
+        records,
+        field_name,
+        lambda old, *a, **k: convert_string_bootstrap_4to5(old),
+    )
+
+
+def _convert_field_bootstrap_4to5_sql(cr, table, field, ids=None):
+    """Convert a field from Bootstrap 4 to 5, using raw SQL queries.
+
+    TODO Support multilang fields.
+
+    :param odoo.sql_db.Cursor cr:
+        Database cursor.
+
+    :param str table:
+        Table name.
+
+    :param str field:
+        Field name, which should contain HTML content.
+
+    :param list ids:
+        List of IDs, to restrict operation to them.
+    """
+    sql = "SELECT id, %s FROM %s " % (field, table)
+    params = ()
+    if ids:
+        sql += "WHERE id IN %s"
+        params = (ids,)
+    cr.execute(sql, params)
+    for id_, old_content in cr.fetchall():
+        new_content = convert_string_bootstrap_4to5(old_content)
+        if old_content != new_content:
+            cr.execute(
+                "UPDATE %s SET %s = %s WHERE id = %s",
+                AsIs(table),
+                AsIs(field),
+                new_content,
+                id_,
+            )
```

Comment:
Is there a bug in the cr.execute() call, assuming syntax is cr.execute(query, params)?
I got the below error during migration which lead me here.

```
2023-11-11 15:54:29,149 58099 CRITICAL test odoo.service.server: Failed to initialize database `test`.
Traceback (most recent call last):
  File "/Users/lijo/work/odoo16/odoo/service/server.py", line 1299, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-16>", line 2, in new
  File "/Users/lijo/work/odoo16/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "/Users/lijo/work/odoo16/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/Users/lijo/work/odoo16/odoo/modules/loading.py", line 523, in load_modules
    migrations.migrate_module(package, 'end')
  File "/Users/lijo/work/oca/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/modules/migration.py", line 18, in migrate_module
    MigrationManager.migrate_module._original_method(self, pkg, stage)
  File "/Users/lijo/work/odoo16/odoo/modules/migration.py", line 189, in migrate_module
    migrate(self.cr, installed_version)
  File "/Users/lijo/miniforge3/envs/odoo16/lib/python3.8/site-packages/openupgradelib/openupgrade.py", line 2277, in wrapped_function
    func(
  File "/Users/lijo/work/oca/OpenUpgrade/openupgrade_scripts/scripts/website/16.0.1.0/end-migration.py", line 81, in migrate
    convert_custom_qweb_templates_bootstrap_4to5(env)
  File "/Users/lijo/work/oca/OpenUpgrade/openupgrade_scripts/scripts/website/16.0.1.0/end-migration.py", line 26, in convert_custom_qweb_templates_bootstrap_4to5
    _convert_field_bootstrap_4to5_sql(env.cr, "ir_ui_view", "arch_db", ids=view_ids)
  File "/Users/lijo/miniforge3/envs/odoo16/lib/python3.8/site-packages/openupgradelib/openupgrade_160.py", line 397, in _convert_field_bootstrap_4to5_sql
    cr.execute(
TypeError: execute() takes from 2 to 4 positional arguments but 6 were given
```

Changing the execute call as below seems to fix it

```
cr.execute(
                "UPDATE %s SET %s = %s WHERE id = %s",
                (AsIs(table), AsIs(field), new_content, id_),
            )

```

Is my understanding wrong?